designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,224 @@
+
+#### Newer NuGet tooling, older build tooling
+
+Say a customer has configured to get latest NuGet in their build tasks. While NuGet.exe and VisualStudio do ship on the same day, the rollout to all build machines is not atomic. The customers could end up in this mismatched situation without an explicit action.
+
+A newer NuGet version *might* generate an assets file that has more/different details than what the build tasks expect.
+A recent example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/vsbuild-msbuild-task-failing-since-2020-11-10/1249432) which is not 100% understood yet.
+
+In this design, we are trying to improve the general mismatched tooling experience, but also consider alternatives for the particular issues customers are hitting with VS 16.8/NuGet 5.8/ SDK 5.0.100.
+
+## Who are the customers
+
+PackageReference customers
+
+## Goals
+
+* Improve the experience for mismatched versions of restore/build tooling.
+
+## Non-Goals
+
+* Introducing explicit backwards/forward compatibility. That's a lateral effort.
+
+## Solution Overview
+
+In the 2 particular problems described above the experience when things go wrong is inconsistent. It's not *always* apparent why something goes wrong.
+This proposal covers improving that warning/error experience.
+
+### NuGet output versioning
+
+When PackageReference restore is run, NuGet writes out the 3 files mentioned above. 2 of those files have a `version`.
+
+* The `project.assets.json` has a version that has historically been used for schema changes, not tooling compatibility checks. For example, aliases addition involved adding more information to the assets file, but not a schema change so this number was not incremented.
+* The `nuget.g.props` contains a `NuGetToolVersion` property.
+
+### Build Tasks warning (SDK)
+
+Given that the implied compatibility exists, the proposal is for the SDK build tasks raise a suppressible warning when the output is not in the expected shape.
+
+* Warn when the restore output is from an older tool than anticipated.
+* Warn when the restore output is from a newer tool than anticipated.
+
+There are 2 technical options and each might have a different effect on the frequency in which this warning is raised.
+
+#### Option 1 - Use assets file version
+
+NuGet increments the `project.assets.json` version every time there's a functional change in behavior. The build tasks in the SDK are built with a specific version of NuGet, and the SDK could raise a warning when the `project.assets.json` version is not the one it expects.
+
+Cons:
+
+* Thas version that has historically been used for schema changes, not tooling compatibility checks. For example, aliases addition involved adding more information to the assets file, but not a schema change so this number was not incremented.
+
+Pros:
+
+* We would only increment this when a functional restore change is made.

```

Comment:
Is a potential pro here that we would _already_ get a warning, for existing past versions of tooling?

Reply:
I think all of the approaches allow us to do that. This one might be the `hardest` to get a warning I think.

1. SDK change would need to look for `4`, since all the previous versions of the assets file wrote `3`. So we need a NuGet + SDK change.
1. SDK change only. The SDK can known the NuGet version it carries and react appropriately.
1. SDK change would need to look at a new property. Alternatively if the parser doesn't see that new property, the SDK could raise a warning. 

I could also be totally misunderstanding you :D

Reply:
No, I think you got it. I was under the impression some kind of warning already existed for bumped version numbers.

Reply:
Ahh...nope, unfortunately :(